### PR TITLE
[Feature] Optimizations for class Qwen3VLMoeVisionModel (Conv3d to Linear) in Qwen3VL

### PR DESCRIPTION
**Pull request by wili-65535**
## Motivation

+ This PR is part of (https://github.com/sgl-project/sglang/issues/18784 and https://github.com/sgl-project/sglang/pull/18559). Corresponding information and performance data is also there.

## Modifications

+ Use `Linear` rather than `Conv3d` to compute the `Qwen3VLVisionPatchEmbed`.
+ Add unit test `test/srt/test_conv3d_to_linear_unittest.py` to ensure the replacement is equivalent.

## Accuracy Tests

1. By unit test `test/srt/test_conv3d_to_linear_unittest.py`.

2. `lm-eval` shows no drop between main branch and this PR, which get the same score:

+ Command:
```bash
lm-eval run --model sglang --model_args pretrained=/workspace/Qwen3-VL-8B-Instruct,dtype=auto,tp_size=1 --tasks gpqa_diamond_zeroshot gpqa_extended_zeroshot gpqa_main_zeroshot gpqa_diamond_n_shot gpqa_extended_n_shot gpqa_main_n_shot
```

+ **before this optimization (main branch baseline)**

| Tasks                  | Version | Filter | n-shot | Metric   |     |  Value |     | Stderr |
| ---------------------- | ------: | ------ | -----: | -------- | --- | -----: | --- | -----: |
| gpqa_diamond_n_shot    |       2 | none   |      0 | acc      | ↑   | 0.3687 | ±   | 0.0344 |
|                        |         | none   |      0 | acc_norm | ↑   | 0.3687 | ±   | 0.0344 |
| gpqa_diamond_zeroshot  |       1 | none   |      0 | acc      | ↑   | 0.3788 | ±   | 0.0346 |
|                        |         | none   |      0 | acc_norm | ↑   | 0.3788 | ±   | 0.0346 |
| gpqa_extended_n_shot   |       2 | none   |      0 | acc      | ↑   | 0.3791 | ±   | 0.0208 |
|                        |         | none   |      0 | acc_norm | ↑   | 0.3791 | ±   | 0.0208 |
| gpqa_extended_zeroshot |       1 | none   |      0 | acc      | ↑   | 0.3773 | ±   | 0.0208 |
|                        |         | none   |      0 | acc_norm | ↑   | 0.3773 | ±   | 0.0208 |
| gpqa_main_n_shot       |       2 | none   |      0 | acc      | ↑   | 0.3862 | ±   | 0.0230 |
|                        |         | none   |      0 | acc_norm | ↑   | 0.3862 | ±   | 0.0230 |
| gpqa_main_zeroshot     |       1 | none   |      0 | acc      | ↑   | 0.4085 | ±   | 0.0232 |
|                        |         | none   |      0 | acc_norm | ↑   | 0.4085 | ±   | 0.0232 |

+ **After the optimization (this PR)**

| Tasks                  | Version | Filter | n-shot | Metric   |     |  Value |     | Stderr |
| ---------------------- | ------: | ------ | -----: | -------- | --- | -----: | --- | -----: |
| gpqa_diamond_n_shot    |       2 | none   |      0 | acc      | ↑   | 0.3687 | ±   | 0.0344 |
|                        |         | none   |      0 | acc_norm | ↑   | 0.3687 | ±   | 0.0344 |
| gpqa_diamond_zeroshot  |       1 | none   |      0 | acc      | ↑   | 0.3788 | ±   | 0.0346 |
|                        |         | none   |      0 | acc_norm | ↑   | 0.3788 | ±   | 0.0346 |
| gpqa_extended_n_shot   |       2 | none   |      0 | acc      | ↑   | 0.3791 | ±   | 0.0208 |
|                        |         | none   |      0 | acc_norm | ↑   | 0.3791 | ±   | 0.0208 |
| gpqa_extended_zeroshot |       1 | none   |      0 | acc      | ↑   | 0.3773 | ±   | 0.0208 |
|                        |         | none   |      0 | acc_norm | ↑   | 0.3773 | ±   | 0.0208 |
| gpqa_main_n_shot       |       2 | none   |      0 | acc      | ↑   | 0.3862 | ±   | 0.0230 |
|                        |         | none   |      0 | acc_norm | ↑   | 0.3862 | ±   | 0.0230 |
| gpqa_main_zeroshot     |       1 | none   |      0 | acc      | ↑   | 0.4085 | ±   | 0.0232 |
|                        |         | none   |      0 | acc_norm | ↑   | 0.4085 | ±   | 0.0232 |

+ `lmms_eval` shows no drop between main branch and this PR, which get the similar score:

+ Command:
```bash
python3 -m lmms_eval --model sglang --model_args model_path=/workspace/Qwen3-VL-8B-Instruct --tasks mmmu_val
```

+ **Before this optimization (main branch baseline)**

```
{'Overall-Art and Design': {'num': 120, 'acc': 0.68333}, 'Art': {'num': 30, 'acc': 0.66667}, 'Art_Theory': {'num': 30, 'acc': 0.9}, 'Design': {'num': 30, 'acc': 0.73333}, 'Music': {'num': 30, 'acc': 0.43333}, 'Overall-Business': {'num': 150, 'acc': 0.40667}, 'Accounting': {'num': 30, 'acc': 0.3}, 'Economics': {'num': 30, 'acc': 0.5}, 'Finance': {'num': 30, 'acc': 0.26667}, 'Manage': {'num': 30, 'acc': 0.5}, 'Marketing': {'num': 30, 'acc': 0.46667}, 'Overall-Science': {'num': 150, 'acc': 0.45333}, 'Biology': {'num': 30, 'acc': 0.53333}, 'Chemistry': {'num': 30, 'acc': 0.36667}, 'Geography': {'num': 30, 'acc': 0.53333}, 'Math': {'num': 30, 'acc': 0.3}, 'Physics': {'num': 30, 'acc': 0.53333}, 'Overall-Health and Medicine': {'num': 150, 'acc': 0.53333}, 'Basic_Medical_Science': {'num': 30, 'acc': 0.66667}, 'Clinical_Medicine': {'num': 30, 'acc': 0.66667}, 'Diagnostics_and_Laboratory_Medicine': {'num': 30, 'acc': 0.36667}, 'Pharmacy': {'num': 30, 'acc': 0.5}, 'Public_Health': {'num': 30, 'acc': 0.46667}, 'Overall-Humanities and Social Science': {'num': 120, 'acc': 0.675}, 'History': {'num': 30, 'acc': 0.6}, 'Literature': {'num': 30, 'acc': 0.83333}, 'Sociology': {'num': 30, 'acc': 0.63333}, 'Psychology': {'num': 30, 'acc': 0.63333}, 'Overall-Tech and Engineering': {'num': 210, 'acc': 0.38095}, 'Agriculture': {'num': 30, 'acc': 0.53333}, 'Architecture_and_Engineering': {'num': 30, 'acc': 0.3}, 'Computer_Science': {'num': 30, 'acc': 0.5}, 'Electronics': {'num': 30, 'acc': 0.3}, 'Energy_and_Power': {'num': 30, 'acc': 0.3}, 'Materials': {'num': 30, 'acc': 0.4}, 'Mechanical_Engineering': {'num': 30, 'acc': 0.33333}, 'Overall': {'num': 900, 'acc': 0.50222}}
2026-03-16T05:00:15.655750+0000 | save_results_aggregated | INFO - Output path not provided, skipping saving results aggregated
sglang (model_path=/workspace/Qwen3-VL-8B-Instruct), gen_kwargs: (), limit: None, offset: 0, num_fewshot: None, batch_size: 1

LMMs-Eval: Probing Intelligence in the Real World
> The unified evaluation toolkit for frontier models.

branch: main
commit: v0.6-72-g88b23e2b

| Tasks  |Filter|n-shot| Metric |   |Value |   |Stderr|
|--------|------|-----:|--------|---|-----:|---|------|
|mmmu_val|none  |     0|mmmu_acc|↑  |0.5022|±  |N/A   |
```

+ **After this optimization (this PR)**

```
{'Overall-Art and Design': {'num': 120, 'acc': 0.68333}, 'Art': {'num': 30, 'acc': 0.66667}, 'Art_Theory': {'num': 30, 'acc': 0.9}, 'Design': {'num': 30, 'acc': 0.73333}, 'Music': {'num': 30, 'acc': 0.43333}, 'Overall-Business': {'num': 150, 'acc': 0.40667}, 'Accounting': {'num': 30, 'acc': 0.3}, 'Economics': {'num': 30, 'acc': 0.5}, 'Finance': {'num': 30, 'acc': 0.26667}, 'Manage': {'num': 30, 'acc': 0.5}, 'Marketing': {'num': 30, 'acc': 0.46667}, 'Overall-Science': {'num': 150, 'acc': 0.45333}, 'Biology': {'num': 30, 'acc': 0.53333}, 'Chemistry': {'num': 30, 'acc': 0.36667}, 'Geography': {'num': 30, 'acc': 0.53333}, 'Math': {'num': 30, 'acc': 0.3}, 'Physics': {'num': 30, 'acc': 0.53333}, 'Overall-Health and Medicine': {'num': 150, 'acc': 0.53333}, 'Basic_Medical_Science': {'num': 30, 'acc': 0.66667}, 'Clinical_Medicine': {'num': 30, 'acc': 0.66667}, 'Diagnostics_and_Laboratory_Medicine': {'num': 30, 'acc': 0.36667}, 'Pharmacy': {'num': 30, 'acc': 0.5}, 'Public_Health': {'num': 30, 'acc': 0.46667}, 'Overall-Humanities and Social Science': {'num': 120, 'acc': 0.675}, 'History': {'num': 30, 'acc': 0.6}, 'Literature': {'num': 30, 'acc': 0.83333}, 'Sociology': {'num': 30, 'acc': 0.63333}, 'Psychology': {'num': 30, 'acc': 0.63333}, 'Overall-Tech and Engineering': {'num': 210, 'acc': 0.38095}, 'Agriculture': {'num': 30, 'acc': 0.53333}, 'Architecture_and_Engineering': {'num': 30, 'acc': 0.3}, 'Computer_Science': {'num': 30, 'acc': 0.5}, 'Electronics': {'num': 30, 'acc': 0.3}, 'Energy_and_Power': {'num': 30, 'acc': 0.3}, 'Materials': {'num': 30, 'acc': 0.4}, 'Mechanical_Engineering': {'num': 30, 'acc': 0.33333}, 'Overall': {'num': 900, 'acc': 0.50222}}
2026-03-16T05:11:01.956747+0000 | save_results_aggregated | INFO - Output path not provided, skipping saving results aggregated
sglang (model_path=/workspace/Qwen3-VL-8B-Instruct), gen_kwargs: (), limit: None, offset: 0, num_fewshot: None, batch_size: 1

LMMs-Eval: Probing Intelligence in the Real World
> The unified evaluation toolkit for frontier models.

branch: main
commit: v0.6-72-g88b23e2b

| Tasks  |Filter|n-shot| Metric |   |Value |   |Stderr|
|--------|------|-----:|--------|---|-----:|---|------|
|mmmu_val|none  |     0|mmmu_acc|↑  |0.5022|±  |N/A   |
```

## Benchmarking and Profiling

+ Part of the performance data is in the original issue, here we file more detailed data.
+ We use H100 GPU to run Qwen3-VL-8B model (the larger the model is, the more significant the performance improvement earns), sending requests with one JPEG image of different resolution
+ Add code below (in `forward()` of `class Qwen3VLMoeVisionModel`) to count the time.

```Python
        # original code
        x = x.to(device=self.device, dtype=self.dtype)

        # timming code
        for _ in range(10):
            y = self.patch_embed(x)
        patch_embed_start = torch.cuda.Event(enable_timing=True)
        patch_embed_end = torch.cuda.Event(enable_timing=True)
        patch_embed_start.record()
        for _ in range(30):
            y = self.patch_embed(x)
        patch_embed_end.record()
        torch.cuda.synchronize(x.device)
        patch_embed_elapsed_ms = patch_embed_start.elapsed_time(patch_embed_end) / 30
        print(f"###[{x.shape[0]}]: {patch_embed_elapsed_ms:.3f} ms")
```

+ The results before / after this PR are shown below. Averagely 6.9x acceleration is earned, and the larger the image is, the better performance is earned.
+ In our old scenario (Qwen3-VL-235B + TP8 + DP1, one request with 20 images of 960x1280), approximately 11.0x acceleration was earned.

| size      | conv3d/ms | linear/ms | SpeedUp |
|:----------|----------:|----------:|--------:|
| 32x32     |     0.122 |     0.070 |    1.74 |
| 64x64     |     0.122 |     0.070 |    1.74 |
| 96x96     |     0.123 |     0.070 |    1.76 |
| 128x128   |     0.123 |     0.071 |    1.73 |
| 160x160   |     0.123 |     0.071 |    1.73 |
| 192x192   |     0.124 |     0.071 |    1.75 |
| 224x224   |     0.129 |     0.071 |    1.82 |
| 256x256   |     0.136 |     0.073 |    1.86 |
| 288x288   |     0.136 |     0.072 |    1.89 |
| 320x320   |     0.175 |     0.070 |    2.50 |
| 352x352   |     0.228 |     0.069 |    3.30 |
| 384x384   |     0.279 |     0.083 |    3.36 |
| 416x416   |     0.261 |     0.076 |    3.43 |
| 448x448   |     0.270 |     0.069 |    3.91 |
| 480x480   |     0.308 |     0.074 |    4.16 |
| 512x512   |     0.334 |     0.069 |    4.84 |
| 544x544   |     0.375 |     0.069 |    5.43 |
| 576x576   |     0.414 |     0.069 |    6.00 |
| 608x608   |     0.499 |     0.070 |    7.13 |
| 640x640   |     0.619 |     0.070 |    8.84 |
| 672x672   |     0.646 |     0.070 |    9.23 |
| 704x704   |     0.663 |     0.069 |    9.61 |
| 736x736   |     0.672 |     0.069 |    9.74 |
| 768x768   |     0.707 |     0.070 |   10.10 |
| 800x800   |     0.747 |     0.068 |   10.99 |
| 832x832   |     0.875 |     0.069 |   12.68 |
| 864x864   |     0.847 |     0.069 |   12.28 |
| 896x896   |     1.028 |     0.069 |   14.90 |
| 928x928   |     1.065 |     0.070 |   15.21 |
| 960x960   |     1.127 |     0.071 |   15.87 |
| 992x992   |     1.174 |     0.071 |   16.54 |
| 1024x1024 |     1.235 |     0.084 |   14.70 |

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/srt/models/qwen3_vl.py [hunk 1]
```diff
@@ -135,20 +135,23 @@ def __init__(self, config) -> None:
...
+            bias=True,
+            dtype=self.proj.weight.dtype,
+        )
+
+    def copy_conv3d_weight_to_linear(self):
+        # Call this after weight loading
+        with torch.no_grad():
+            self.linear.weight.copy_(self.proj.weight.view(self.embed_dim, -1))
+            self.linear.bias.copy_(self.proj.bias)
+        del self.proj
```
[the remark this answers is not in the document]
Comment (wili-65535) [reply python/sglang/srt/models/qwen3_vl.py:163]: Fixed.

**Review by yuan-luo — APPROVED** [review]

**Comment by yuan-luo** [comment]
> Hi @yuan-luo , I see your [PR20033](https://github.com/sgl-project/sglang/pull/20033) for GLM4v has been merged. Cheers! I've adjusted this PR aligned with yours. Now I wonder,
> 
> * How can we move this PR forward? Does our CI pipeline resume to work?
> * Are the accuracy and performance results in the description enough for this PR?
> * Do we still need the unit test [TestConv3dToLinear](test/srt/test_conv3d_to_linear_unittest.py) beside your [patch_embed_linear_conv3d](test/registered/vlm/test_patch_embed_perf.py)?
> 
> Oh, I see the CI is rerunning, thank you...

@wili-65535 We can probably keep this unit test as it is related with Qwen3VLMoE. BTW, can we move test to the vlm folder?

Let's wait for the CI passed.

**Comment by wili-65535** [comment]
> The latest CI results seem to indicate that this change may have introduced an accuracy issue. Could you please take a look when you have a chance? Many thanks! https://github.com/sgl-project/sglang/actions/runs/22891107783/job/66415614526?pr=19788

Fixed, root cause is Qwen3Omni needs adjust, too.
